qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -11,35 +11,119 @@
 # that they have been altered from the originals.
 
 
-"""Synthesize high-level objects."""
+"""Synthesize higher-level objects."""
+
+import stevedore
 
 from qiskit.converters import circuit_to_dag
 from qiskit.transpiler.basepasses import TransformationPass
 from qiskit.dagcircuit.dagcircuit import DAGCircuit
-from qiskit.quantum_info.synthesis.clifford_decompose import decompose_clifford
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+# ToDo: possibly choose a better name for this data structure.
+class HLSConfig:
+    """
+    To each higher-level-object (e.g., "clifford", "linear_function", etc.) we associate a
+    pair consisting of the name of a synthesis algorithm and additional arguments it requires.
+    There are two special values for the name:
+        "default" is the default synthesis algorithm (chosen by Qiskit developers)
+        "none" is to skip synthesizing this object
+    All other names should be declared in entry points in setup.
+    """
+
+    def __init__(self, default_or_none: True):
+        """
+        Creates a high-level-synthesis config.
+        Args:
+            default_or_none: specifies the default method for every higher-level object,
+                either "default" (use the "default" synthesize method if available),
+                or "none" (skip synthesizing)
+        """
+        self.default_or_none = default_or_none
+        self.methods = dict()
```

Comment:
These are good suggestions. Based on the slack discussion, for each higher-level-object `HLSConfig` now specifies a _list_ of synthesis methods and their arguments. I have changed `default_or_none` to `use_default_on_unspecified`. Finally, I have changed `HLSConfig.__init__` to be of the form 
```
def __init__(self, use_default_on_unspecified: True, **kwargs):
    ...
```
This now allows to specify all relevant information in `__init__`, yet it still looks a bit too verbose. For instance,
```
hls_config = HLSConfig(use_default_on_unspecified=True,
                       linear_function=[("depth_opt", {}), ("count_opt", {})],
                       clifford=[])
qct = transpile(qc, optimization_level=1, hls_config=hls_config)
```
says to synthesize `LinearFunctions` using `depth_opt` (with no additional arguments), and if this fails -- then with `count_opt` (with no additional arguments), and to skip synthesis for `Cliffords` (we can now pass an empty list instead of the `none` method). What do you think about the current syntax?

Reply:
Thanks, I think this looks more consistent with elsewhere in Qiskit. I agree that overall this looks verbose, but this seems okay since this won't be the long-term API most users see, only those users who want to specify (or override) exactly the synthesis approaches used by `HighLevelSynthesis`, when all of this becomes necessary information.

The only piece I could foresee being simpler is allowing users to omit the `{}` in the case where the synthesis method takes no parameters, but this seems rare to me (and we can return to add it later) so I'm okay leaving it as it for now.